includes/class-amp-theme-support.php

```diff
@@ -2183,17 +2183,22 @@ public static function prepare_response( $response, $args = [] ) {
 
 		$dom = Document::fromHtml( $response );
 
-		// Move anything after </html>, such as Query Monitor output added at shutdown, to be moved before </body>.
-		while ( $dom->documentElement->nextSibling ) {
+		/*
+		 * Move any non-comment elements after </html>, such as Query Monitor output added at shutdown, to be moved
+		 * before </body>.
+		 */
+		$next_sibling = $dom->documentElement->nextSibling;
+		while ( $next_sibling ) {
 			// Trailing elements after </html> will get wrapped in additional <html> elements.
-			if ( 'html' === $dom->documentElement->nextSibling->nodeName ) {
-				while ( $dom->documentElement->nextSibling->firstChild ) {
-					$dom->body->appendChild( $dom->documentElement->nextSibling->firstChild );
+			if ( 'html' === $next_sibling->nodeName ) {
+				while ( $next_sibling->firstChild ) {
+					$dom->body->appendChild( $next_sibling->firstChild );
 				}
-				$dom->removeChild( $dom->documentElement->nextSibling );
-			} else {
-				$dom->body->appendChild( $dom->documentElement->nextSibling );
+				$dom->removeChild( $next_sibling );
+			} elseif ( ! $next_sibling instanceof DOMComment ) {
+				$dom->body->appendChild( $next_sibling );
 			}
+			$next_sibling = $next_sibling->nextSibling;
```

Comment:
This turns out to not be quite right. If element are moved but not comment nodes, then the source stacks can get confused.

Consider a `footer.php` ending with:

```php
		<?php wp_footer(); ?>

	</body>
	<?php do_action( 'after_closing_body' ); ?>
</html>
<?php do_action( 'after_closing_html' ); ?>
```

And a plugin like this:

```php
<?php
/**
 * Plugin Name: Try After Actions
 */

add_action( 'after_closing_body', function () {
	echo '<script>after_closing_body</script>';
} );

add_action( 'after_closing_html', function () {
	echo '<script>after_closing_html</script>';
} );
```

At the moment this results in validation errors looking like this:

![image](https://user-images.githubusercontent.com/134745/76151377-6628bd80-6069-11ea-82e1-145fa0fa10c5.png)

Notice the lack of source identified for the first.

So what I think needs to be done is _every_ node appearing after `</body>` and `</html>` should be moved to be appended to the `body`, regardless of the node type. This will ensure that the source stack comments maintain their relative position with the elements they annotate.

Ultimately it should rather go into `\AmpProject\Dom\Document::loadHTML()`

Reply:
Patch incoming…

Reply:
See e4f10d1 and 62fd552